File: Leetcode/538.py
# 538. 把二叉搜索树转换为累加树    dfs
# link：https://leetcode-cn.com/problems/convert-bst-to-greater-tree/

class Solution:
    def __init__(self) -> None:
        self.sum = 0

    def convertBST(self, root: TreeNode) -> TreeNode:
        if not root:
            return None

        self.convertBST(root.right)
        self.sum += root.val
        root.val = self.sum
        self.convertBST(root.left)

        return root

File: Leetcode/538.cpp
/**
 * 538. 把二叉搜索树转换为累加树    dfs
 * link：https://leetcode-cn.com/problems/convert-bst-to-greater-tree/
 */

class Solution
{
    int sum = 0;

public:
    TreeNode *convertBST(TreeNode *root)
    {
        // if (root == nullptr)
        //     return nullptr;

        // convertBST(root->right);
        // sum += root->val;
        // root->val = sum;
        // convertBST(root->left);

        convert(root);
        return root;
    }

    //without return value
    void convert(TreeNode *cur)
    {
        if (cur == nullptr)
            return;
        convert(cur->right);
        sum += cur->val;
        cur->val = sum;
        convert(cur->left);
    }
};
